exonum/src/storage/sparse_list_index.rs

```diff
@@ -0,0 +1,838 @@
+
+    /// Removes the element with the given index from the list and returns it,
+    /// or None if it is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert_eq!(0, index.capacity());
+    ///
+    /// index.push(10);
+    /// index.push(12);
+    ///
+    /// assert_eq!(Some(10), index.remove(0));
+    /// assert_eq!(None, index.remove(0));
+    /// assert_eq!(2, index.capacity());
+    /// assert_eq!(1, index.len());
+
+    /// assert_eq!(Some(12), index.remove(1));
+    /// assert_eq!(2, index.capacity());
+    /// ```
+    pub fn remove(&mut self, index: u64) -> Option<V> {
+        let mut size = self.size();
+        if index >= size.capacity {
+            warn!(
+                "Index out of bounds: \
+                    the capacity is {} but the index is {}",
+                size.capacity,
+                index
+            );
+            return None;
+        }
+        let v = self.base.get(&index);
+        if v.is_some() {
+            self.base.remove(&index);
+            size.length -= 1;
+            self.set_size(size);
+        }
+        v
+    }
+
+    /// Extends the list with the contents of an iterator.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert!(index.is_empty());
+    ///
+    /// index.extend([1, 2, 3].iter().cloned());
+    /// assert_eq!(3, index.capacity());
+    /// ```
+    pub fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = V>,
+    {
+        let mut size = self.size();
+        for value in iter {
+            self.base.put(&size.capacity, value);
+            size.capacity += 1;
+            size.length += 1;
+        }
+        self.set_size(size);
+    }
+
+    /// Changes a value at the specified position. If the position contains empty value it
+    /// also increments elements count. If a value of the index of new element is greater than
+    /// current capacity, the capacity of the list considered index + 1 and all further elements
+    /// without specific index value will be appended after this index.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.push(1);
+    /// assert_eq!(Some(1), index.get(0));
+    ///
+    /// index.set(0, 10);
+    /// assert_eq!(Some(10), index.get(0));
+    /// ```
+    pub fn set(&mut self, index: u64, value: V) {
```

Comment:
Since you still need to get the old value by the index, it may be useful to return it from the method as `-> Option<V>`